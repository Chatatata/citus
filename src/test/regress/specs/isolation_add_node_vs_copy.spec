setup
{
	truncate pg_dist_shard_placement;
	truncate pg_dist_shard;
	truncate pg_dist_partition;
	truncate pg_dist_colocation;
	truncate pg_dist_node;
	
	SELECT master_add_node('localhost', 57637);
	
    CREATE TABLE test_reference_table (test_id integer);
    SELECT create_reference_table('test_reference_table');
}

teardown
{
    DROP TABLE IF EXISTS test_reference_table CASCADE;

	SELECT master_add_node('localhost', 57637);
	SELECT master_add_node('localhost', 57638);
}

session "s1"

step "s1-begin"
{
    BEGIN;
}

step "s1-add-second-worker"
{
   	SELECT master_add_node('localhost', 57638);
}

step "s1-remove-second-worker"
{
   	SELECT master_remove_node('localhost', 57638);
}

step "s1-commit"
{
    COMMIT;
}

session "s2"

step "s2-begin"
{
    BEGIN;
}

step "s2-copy-to-reference-table" 
{
	COPY test_reference_table FROM PROGRAM 'echo "1\n2\n3\n4\n5"';
}

step "s2-commit"
{
    COMMIT;
}

step "s2-print-content"
{
	SELECT run_command_on_placements('test_reference_table', 'select count(*) from %s');
}

# verify that copy gets the invalidation and re-builts its metadata cache
# note that we need to run the same test twice to ensure that metadata is cached
# otherwise the test would be useless since the cache would be empty and the 
# metadata data is gathered from the tables directly
permutation "s1-begin" "s1-add-second-worker" "s2-begin" "s2-copy-to-reference-table" "s1-commit" "s2-commit" "s2-print-content" "s1-remove-second-worker" "s1-begin" "s1-add-second-worker" "s2-begin" "s2-copy-to-reference-table" "s1-commit" "s2-commit" "s2-print-content"
