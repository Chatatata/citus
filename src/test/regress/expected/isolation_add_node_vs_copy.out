Parsed test spec with 2 sessions

starting permutation: s1-begin s1-add-second-worker s2-begin s2-copy-to-reference-table s1-commit s2-commit s2-print-content s1-remove-second-worker s1-begin s1-add-second-worker s2-begin s2-copy-to-reference-table s1-commit s2-commit s2-print-content
create_reference_table

               
step s1-begin: 
    BEGIN;

step s1-add-second-worker: 
   	SELECT master_add_node('localhost', 57638);

master_add_node

(4,4,localhost,57638,default,f,t)
step s2-begin: 
    BEGIN;

step s2-copy-to-reference-table: 
	COPY test_reference_table FROM PROGRAM 'echo "1\n2\n3\n4\n5"';
 <waiting ...>
step s1-commit: 
    COMMIT;

step s2-copy-to-reference-table: <... completed>
step s2-commit: 
    COMMIT;

step s2-print-content: 
	SELECT run_command_on_placements('test_reference_table', 'select count(*) from %s');

run_command_on_placements

(localhost,57637,102197,t,5)
(localhost,57638,102197,t,5)
step s1-remove-second-worker: 
   	SELECT master_remove_node('localhost', 57638);

master_remove_node

               
step s1-begin: 
    BEGIN;

step s1-add-second-worker: 
   	SELECT master_add_node('localhost', 57638);

master_add_node

(5,5,localhost,57638,default,f,t)
step s2-begin: 
    BEGIN;

step s2-copy-to-reference-table: 
	COPY test_reference_table FROM PROGRAM 'echo "1\n2\n3\n4\n5"';
 <waiting ...>
step s1-commit: 
    COMMIT;

step s2-copy-to-reference-table: <... completed>
step s2-commit: 
    COMMIT;

step s2-print-content: 
	SELECT run_command_on_placements('test_reference_table', 'select count(*) from %s');

run_command_on_placements

(localhost,57637,102197,t,10)
(localhost,57638,102197,t,10)
master_add_node

(5,5,localhost,57638,default,f,t)
